File: Gombang/Weekly Contest 404/3200_Maximum_Height_of_a_Triangle.cs
public class Solution
{
    public int MaxHeightOfTriangle(int red, int blue)
    {
        // red를 먼저 쌓았을 때의 높이와 blue를 먼저 쌓았을 때의 높이 중에 더 큰 높이를 반환.
        return Math.Max(GetHeight(red, blue), GetHeight(blue, red));
    }

    private int GetHeight(int ball1, int ball2)
    {
        int height = 0;

        int i = 1;
        while (true)
        {
            // 홀수 층 쌓을때 필요한 공의 개수 1, 3, 5, 7, ...
            ball1 -= (2 * i) - 1;
            if (ball1 < 0)
                break;

            height++;

            // 짝수 층 쌓을때 필요한 공의 개수 2, 4, 6, 8, ...
            ball2 -= (2 * i);
            if (ball2 < 0)
                break;

            height++;
            i++;
        }

        return height;
    }
}

File: Lucy/Weekly Contest 404/3200_Maximum_Height_of_a_Triangle.js
/**
 * 3200. Maximum Height of a Triangle
 * URL: https://leetcode.com/problems/maximum-height-of-a-triangle/
 * TOPIC:
 * DIFFICULTY: EASY
 * DATE: 2024.07.02(TUE)~
 */

/**
 * @param {number} red
 * @param {number} blue
 * @return {number}
 */
var maxHeightOfTriangle = function (red, blue) {
  const calcHeightOfTriangle = (startColor, red, blue) => {
    const isColorPossible = (color, row) => color >= row;

    let currColor = startColor;
    let height = 0;

    for (
      let row = 1;
      (currColor === "r" && isColorPossible(red, row)) ||
      (currColor === "b" && isColorPossible(blue, row));
      row++
    ) {
      if (currColor === "r") {
        currColor = "b";
        red -= row;
      } else if (currColor === "b") {
        currColor = "r";
        blue -= row;
      }
      height = row;
    }
    return height;
  };
  // red가 top인 경우: red - blue - red - blue
  const resOfRedTop = calcHeightOfTriangle("r", red, blue);

  // blue가 top인 경우
  const resOfBlueTop = calcHeightOfTriangle("b", red, blue);

  return Math.max(resOfRedTop, resOfBlueTop);
};

console.log(maxHeightOfTriangle(10, 1));
